Change TaskInstance and TaskReschedule PK from execution_date to run_id

ashb opened the pull request:
Since TaskReschedule had an existing FK to TaskInstance we had to move
change both of these at the same time.

This puts an explicit FK constraint between TaskInstance and DagRun,
meaning that we can remove a lot of "find TIs without DagRun" code in
the scheduler too, as that is no longer a possible situation.

Since there is now an explicit foreign key between TaskInstance and
DagRun, we can remove a lot of the "cleanup" code in the scheduler that
was dealing with this.

This PR is, sadly, unavoidably large as it changes a fundamental part of the Airflow system, the primary key of the TaskInstance table :(

Closes  #17046

Related #16302, #17030


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb committed:
Change TaskInstance and TaskReschedule from execution_date to run_id

Since TaskReschedule had an existing FK to TaskInstance we had to move
change both of these at the same time.

This puts an explicit FK constraint between TaskInstance and DagRun,
meaning that we can remove a lot of "find TIs without DagRun" code in
the scheduler too, as that is no longer a possible situation.

Since there is now an explicit foreign key between TaskInstance and
DagRun, we can remove a lot of the "cleanup" code in the scheduler that
was dealing with this.

uranusjr commented on airflow/cli/commands/task_command.py:
```diff
@@ -51,15 +51,17 @@
+            dag_run.session.query(DagRun).filter(
+                DagRun.dag_id == task.dag_id,
+                DagRun.execution_date == execution_date,
+            ).one()
```
```suggestion
            session.query(DagRun).filter(
                DagRun.dag_id == task.dag_id,
                DagRun.execution_date == execution_date,
            ).one()
```

?

ashb replied:
Hmmmm I _thought_ I fixed that already :/ Clearly not.

uranusjr commented on airflow/cli/commands/task_command.py:
```diff
@@ -51,15 +51,17 @@
+                DagRun.dag_id == task.dag_id,
+                DagRun.execution_date == execution_date,
+            ).one()
         except (ParserError, TypeError):
```
Can these still happen or should we change this to `NoResultFound`?

ashb replied:
These can happen from the `timezone.parse` on L60, but yes we should add NoResultFound too

uranusjr commented on airflow/jobs/backfill_job.py:
```diff
@@ -645,16 +645,14 @@ def tabulate_ti_keys_set(set_ti_keys: Set[TaskInstanceKey]) -> str:
             sorted_ti_keys = sorted(
                 set_ti_keys,
-                key=lambda ti_key: (ti_key.execution_date, ti_key.dag_id, ti_key.task_id, ti_key.try_number),
+                key=lambda ti_key: (ti_key.run_id, ti_key.dag_id, ti_key.task_id, ti_key.try_number),
```
Since we have a fk constraint and relationship, can we do some join and keep `TaskInstanceKey.execution_date` to be backed by the DagRun’s `execution_date` instead? (i.e. don’t change this function, but change how `TaskInstanceKey` is created.) The current change would break ordering if a DagRun has a custom run ID.

ashb replied:
This would change the order that errors are printed only so it's not "significant" to anything.

(The session handling here is all over the shop, and if the TI is detached from the session and hasn't yet loaded `ti.dag_run` then you get a DetachedObject exception)

ashb replied:
https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html?highlight=joinedload#sqlalchemy.orm.contains_eager may be useful in various places (`joinedload` would join against the table _twice_)

kaxil replied:
in that case L645 needs to be updated

ashb replied:
Already updated in follow on pr

uranusjr commented on airflow/jobs/backfill_job.py:
```diff
@@ -645,16 +645,14 @@ def tabulate_ti_keys_set(set_ti_keys: Set[TaskInstanceKey]) -> str:
-                set_tis, key=lambda ti: (ti.execution_date, ti.dag_id, ti.task_id, ti.try_number)
-            )
-            tis_values = ((ti.dag_id, ti.task_id, ti.execution_date, ti.try_number) for ti in sorted_tis)
+            sorted_tis = sorted(set_tis, key=lambda ti: (ti.run_id, ti.dag_id, ti.task_id, ti.try_number))
+            tis_values = ((ti.dag_id, ti.task_id, ti.run_id, ti.try_number) for ti in sorted_tis)
```
Similar to above, but here we need to do a join when these `TaskInstance` objects were fetched so `ti.dag_run.execution_date` can be cheap.

kaxil replied:
same as above L653 needs to be updated if we sort by run_id

kaxil replied:
Would be good to keep ordering by `execution_date` or even `start_date`, run_id feels odds for sorting here

ephraimbuddy reviewed (commented):
I'm wondering if https://github.com/apache/airflow/blob/4da4c186ecdcdae308fe8b4a7994c21faf42bc96/airflow/www/views.py#L1485-L1486 need a change too?

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb commented on tests/core/test_core.py:
```diff
@@ -314,64 +330,12 @@ def test_task_get_template(self):
             assert value == expected_value
             assert [str(m.message) for m in recorder] == [message]
 
-    def test_local_task_job(self):
```
These tests are duplicating what was already in test_local_task_job -- we don't need them again.

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -189,80 +189,6 @@ def is_alive(self, grace_multiplier: Optional[float] = None) -> bool:
         )
 
-    @provide_session
-    def _change_state_for_tis_without_dagrun(
```
This is deleted, as the DB won't let us have TIs without Dagruns anymore :)

ephraimbuddy replied:
Huge, can clear the issue of tasks getting stuck!

ephraimbuddy replied:
Since tasks will no longer exist without dagrun, obviously the dag will fail or succeed and tis won't be stuck. I think we should remove these lines.:
https://github.com/apache/airflow/blob/5d90bcb93a4318de7db7fd5527bb45c823fb0dab/airflow/jobs/scheduler_job.py#L603-L604

ashb replied:
I think this can happen if (for example) there is a crash in the worker early on in the "boot" process, so it can't report it's own status in the DB.

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -321,12 +247,12 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
         query = (
             session.query(TI)
             .outerjoin(TI.dag_run)
```
```suggestion
            .join(TI.dag_run)
```

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1257,24 +1257,46 @@ def get_flat_relatives(self, upstream: bool = False):
-            TaskInstance(self, info.logical_date).run(
+            try:
+                ti = TaskInstance(self, info.logical_date)
+            except DagRunNotFound:
```
It’s pretty surprising for a class init to throw an exception. I understand it’s unavoidable if `logical_date` is passed in in general, but perhaps her we can avoid doing it by getting the `DagRun` instance beforehand instead of passing in `info.logical_date`? We can also bulk-read the `run_id`s before the loop instead of reading them one by one.

Something like

```python
logical_dates = [
    info.logical_date
    for info in self.dag.iter_dagrun_infos_between(start_date, end_date, align=False)
]
dagruns = {
    run.execution_date: run
    for run in session.query(DagRun).filter(DagRun.execution_date.in_(logical_dates))
}
for date in logical_dates:
    try:
        run = dagruns[date]
    except KeyError:
        # Create a "fake" run…
    ti = TaskInstance(self, run_id=run.run_id)
```

ashb replied:
(`# Create a "fake" run…` -- it can't be a fake one, it needs to be a real one in the DB as the ti gets committed by `ti.run()`

uranusjr replied:
You’re right; by “fake” I meant “created manually to match the supposed internal db state”, sorry for the lax choice of word.

ashb replied:
https://github.com/apache/airflow/commit/2a7f8a411ee43b0ee006e3fb046f559c2087b1d9#diff-848f325ace55b3504e8052fecdb53c0f295c891b67a6d90e9341cbe79cc545fbR1281

Ah I thought you meant FakeDagRun like we had previously in get_template_context.

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -675,17 +671,15 @@ def verify_integrity(self, session: Session = None):
-            self.log.info(
-                'Hit IntegrityError while creating the TIs for ' f'{dag.dag_id} - {self.execution_date}.'
-            )
+            self.log.info('Hit IntegrityError while creating the TIs for ' f'{dag.dag_id} - {self.run_id}.')
```
```suggestion
            self.log.info('Hit IntegrityError while creating the TIs for %s - %s.', dag.dag_id, self.run_id)
```

?

ashb replied:
2a7f8a411

(I just changed the values, didn't look that closely at what it was previously doing.)

uranusjr commented on airflow/models/skipmixin.py:
```diff
@@ -37,36 +38,23 @@
     """A Mixin to skip Tasks Instances"""
 
-    def _set_state_to_skipped(self, dag_run, execution_date, tasks, session):
+    def _set_state_to_skipped(self, dag_run, tasks, session):
```
```suggestion
    def _set_state_to_skipped(self, dag_run: DagRun, tasks: Iterable[BaseOperator], session: Session):
```

This might help catch lingering usages still not passing in a DagRun… (Also need to annotate `skip` below)

ashb replied:
2a7f8a411

uranusjr commented on airflow/models/skipmixin.py:
```diff
@@ -91,7 +79,28 @@ def skip(
             return
 
-        self._set_state_to_skipped(dag_run, execution_date, tasks, session)
+        if execution_date and not dag_run:
+            from airflow.models.dagrun import DagRun
+
+            warnings.warn(
+                "Passing an execution_date to `skip()` is deprecated in favour of passing a dag_run",
+                DeprecationWarning,
+                stacklevel=2,
+            )
```
I think we also need a warning when `dag_run` and `execution_date` are *both* passed to say `execution_date` is now ignored and should be removed. Or maybe we can check the value matches `dagrun.execution_date` and raise a `ValueError`. Or both of these.

ashb replied:
2a7f8a411

uranusjr reviewed (commented):
Only read about 1/5 (probably less) of the patch. Generally seems fine to me but some thoughts on the implementation.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -1723,30 +1736,22 @@ def is_eligible_to_retry(self):
+            if task.dag.params:
+                params.update(task.dag.params)
 
         params = {}  # type: Dict[str, Any]
```
The first `params` is shadowed. Accidental?

ashb replied:
Yes, _very_.

ashb replied:
I think this was a bad conflict resolution.

ashb replied:
2a7f8a411

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -389,27 +399,53 @@ class TaskInstance(Base, LoggingMixin):
+                        .filter_by(dag_id=self.dag_id, execution_date=execution_date)
+                        .one()
+                    )
+                except NoResultFound:
```
```suggestion
                run_id = (
                    session.query(DagRun.run_id)
                    .filter_by(dag_id=self.dag_id, execution_date=execution_date)
                    .scalar()
                )
                if run_id is None:
```

ashb replied:
Oh yes that's nicer.

ashb replied:
2a7f8a411

uranusjr commented on docs/apache-airflow/migrations-ref.rst:
```diff
@@ -23,6 +23,8 @@ Here's the list of all the Database Migrations that are executed via when you ru
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
+| ``7b2661a43ba3`` (head)        | ``142555e44c17`` |                 | Change TaskInstance and TaskReschedule tables from execution_date to run_id.          |
++--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
 | ``142555e44c17`` (head)        | ``54bebd308c5f`` |                 | Add ``data_interval_start`` and ``data_interval_end`` to ``DagRun``                   |
```
```suggestion
| ``142555e44c17``               | ``54bebd308c5f`` |                 | Add ``data_interval_start`` and ``data_interval_end`` to ``DagRun``                   |
```

ashb replied:
2a7f8a411

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -1092,13 +1102,16 @@ def get_dagrun(self, session: Session = None):
+        dr = session.query(DagRun).filter(DagRun.dag_id == self.dag_id, DagRun.run_id == self.run_id).one()
+
+        # Record it in the instance for next timme. This means that `self.execution_date` will work correctly
+        set_committed_value(self, 'dag_run', dr)
```
```suggestion
        # Record it in the instance for next time. This means that `self.execution_date` will work correctly
        set_committed_value(self, 'dag_run', dr)
```

This also means `info.attrs.dag_run` will become loaded next time this is called, right?

ashb replied:
Yes.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -2241,8 +2247,10 @@ def xcom_pull(
             dag_id = self.dag_id
 
+        execution_date = self.get_dagrun(session).execution_date
+
         query = XCom.get_many(
-            execution_date=self.execution_date,
+            execution_date=execution_date,
```
Backref-ing #16370 so we remember to change this when XCom uses run_id.

uranusjr commented on airflow/providers/google/cloud/operators/bigquery.py:
```diff
@@ -60,8 +61,12 @@ class BigQueryConsoleLink(BaseOperatorLink):
+            task_ids=operator.task_id,
+            execution_date=dttm,
+            key='job_id',
+        )
```
Also #16370. Should we deprecate using `execution_date` in `XCom.get_one()` and `XCom.get_many()`? If we do, how do we keep backward compatibility here while avoid emitting the DeprecationWarning?

uranusjr commented on airflow/ti_deps/deps/dagrun_exists_dep.py:
```diff
@@ -29,27 +29,9 @@ class DagrunRunningDep(BaseTIDep):
+                "the state '{}'.".format(dr.state)
             )
-
-            if len(running_dagruns) >= dag.max_active_runs:
-                reason = (
-                    "The maximum number of active dag runs ({}) for this task "
-                    "instance's DAG '{}' has been reached.".format(dag.max_active_runs, ti.dag_id)
-                )
-            else:
-                reason = "Unknown reason"
-            yield self._failing_status(reason=f"Task instance's dagrun did not exist: {reason}.")
```
Why is it OK to drop this logic?

ashb replied:
This was only triggered for the case when there was no dag run, which is not possible now.

(And the scheduler now manages the number of active DagRuns)

uranusjr commented on airflow/ti_deps/dep_context.py:
```diff
@@ -85,23 +84,16 @@ def __init__(
         self.finished_tasks = finished_tasks
 
-    def ensure_finished_tasks(self, dag, execution_date: pendulum.DateTime, session: Session):
+    def ensure_finished_tasks(self, dag_run, session: Session):
```
```suggestion
    def ensure_finished_tasks(self, dag_run: "DagRun", session: Session):
```

(And an import of course.)

uranusjr commented on airflow/www/utils.py:
```diff
@@ -473,6 +474,12 @@ def is_extendedjson(self, col_name):
             )
         return False
 
+    def get_col_default(self, col_name: str) -> Any:
+        if col_name not in self.list_columns:
+            # Handle AssociationProxy etc, or anything that isn't a "real" column
+            return None
+        return super().get_col_default(col_name)
```
I’m not sure what to think about this (not your code but the fact we need this block of code in the first place).

ashb replied:
Quite.

uranusjr commented on airflow/www/views.py:
```diff
@@ -1096,7 +1097,8 @@ def rendered_k8s(self):
         task = dag.get_task(task_id)
-        ti = models.TaskInstance(task=task, execution_date=dttm)
+        dag_run = dag.get_dagrun(execution_date=execution_date)
+        ti = dag_run.get_task_instance(task_id=task.task_id)
```
At some point we should change these views to use `run_id` right?

ashb replied:
Yes, that involves changing all the URLs, and that felt like it should be another PR as this one is big enough already :)

uranusjr commented on tests/executors/test_celery_executor.py:
```diff
@@ -324,11 +322,11 @@ def test_try_adopt_task_instances(self):
             task_2 = BaseOperator(task_id="task_2", start_date=start_date)
 
-        ti1 = TaskInstance(task=task_1, execution_date=exec_date)
+        ti1 = TaskInstance(task=task_1, run_id=None)
```
I feel we should create a real DagRun here that uses `exec_date` instead.

uranusjr commented on tests/executors/test_celery_executor.py:
```diff
@@ -363,8 +360,8 @@ def test_check_for_stalled_adopted_tasks(self):
-        key_1 = TaskInstanceKey(dag.dag_id, task_1.task_id, exec_date, try_number)
-        key_2 = TaskInstanceKey(dag.dag_id, task_2.task_id, exec_date, try_number)
+        key_1 = TaskInstanceKey(dag.dag_id, task_1.task_id, "runid", try_number)
+        key_2 = TaskInstanceKey(dag.dag_id, task_2.task_id, "runid", try_number)
```
Less sure about this; is creating real DagRun and TaskInstance objects worthwhile?

ashb replied:
Might be worth it, yes. Two reasons I did this: it was late, and nothing in the Executor cares about DagRun or strictly speaking executor_date.


Thoughts?

uranusjr commented on tests/jobs/test_triggerer_job.py:
```diff
@@ -305,22 +303,14 @@ def test_invalid_trigger(session):
+    with dag_maker(dag_id='test_invalid_trigger'):
+        DummyOperator(task_id='dummy1')
 
+    dr = dag_maker.create_dagrun()
```
Do we need to pass `session` in?

ashb replied:
Done in 2a7f8a411 (to the dag_maker call)

uranusjr reviewed (commented):
Still about 1/4 to go (all in `tests`), chipping away…

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

kaxil commented on airflow/jobs/backfill_job.py:
```diff
@@ -865,13 +863,7 @@ def reset_state_for_orphaned_tasks(self, filter_by_dag_run=None, session=None):
-                        TaskInstance.execution_date == DagRun.execution_date,
-                    ),
-                )
+                .join(TaskInstance.dag_run)
```
Worth using `contains_eager` here too?

ashb replied:
Yes probably -- didn't know about that when I first started.

kaxil commented on airflow/jobs/scheduler_job.py:
```diff
@@ -321,12 +247,12 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session =
             session.query(TI)
             .outerjoin(TI.dag_run)
-            .filter(or_(DR.run_id.is_(None), DR.run_type != DagRunType.BACKFILL_JOB))
+            .filter(DR.run_type != DagRunType.BACKFILL_JOB)
```
`contains_eager` migth help here too if I understand it correct

kaxil commented on airflow/migrations/versions/7b2661a43ba3_taskinstance_keyed_to_dagrun.py:
```diff
@@ -0,0 +1,211 @@
+        batch_op.alter_column('run_id', existing_type=run_id_col_type, existing_nullable=True, nullable=False)
+
+        # TODO: Is this right for non-postgres?
+        batch_op.drop_constraint('task_instance_pkey', type_='primary')
```
This constraint wasn't named so would be different for different database so we might have to do something like https://github.com/apache/airflow/blob/main/airflow/migrations/versions/bbf4a7ad0465_remove_id_column_from_xcom.py#L75-L80

kaxil replied:
https://alembic.sqlalchemy.org/en/latest/naming.html

kaxil reviewed (commented):
Have reviewed partially (just few files)

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

uranusjr committed:
Convert TestLogView to pytest and create_dummy_dag

ashb committed:
fixup! Change TaskInstance and TaskReschedule from execution_date to run_id

uranusjr committed:
Fix dag file processor manager tests

uranusjr committed:
Partially fix TestDagFileProcessor

There are still a couple of tests failing, especially one with
IntegrationError I can't discern. Leaving this for now.

uranusjr committed:
Fix lineage tests

uranusjr committed:
Fix subdag operator tests

uranusjr committed:
Fix Python operator tests

uranusjr committed:
Fix LatestOnlyOperator tests

uranusjr committed:
Change smart sensor lookup to base on run_id

ashb committed:
Fix base_sensor tests

uranusjr committed:
Fix DagRun view tests

ashb committed:
Fix AWS provider tests

uranusjr committed:
Rewrite to fix Druid provider tests

uranusjr committed:
Fix Hive operator tests

uranusjr committed:
Fix Kylin and Spark providers tests

uranusjr committed:
Fix Kubernetes provider tests

uranusjr committed:
Fix base executor tests

uranusjr committed:
Get DagRun from TI by run_id, not execution_date

uranusjr committed:
Fix DAG tests - Relax DagRun loaded value check

uranusjr committed:
Don't overwrite data_interval provided in test cases

uranusjr committed:
Fix DagRun tests

uranusjr committed:
TI.refresh_from_db() new emits fewer queries, yay!

uranusjr committed:
Fix DagrunRunningDep tests

uranusjr committed:
Delete DagrunIdDep test that's no longer possible

uranusjr committed:
Fix NotPreviouslySkippedDep tests

uranusjr committed:
Fix RunnableExecDateDep tests

uranusjr committed:
Fix typo in trigger rule deps test

uranusjr committed:
Fix dot renderer tests

uranusjr committed:
Fix utils helper tests

uranusjr committed:
Fix log handler tests

uranusjr committed:
Fix custom task log formatter tests

uranusjr committed:
Fix log reader tests

uranusjr committed:
Fix GCS task handler tests

uranusjr committed:
Fix Stackdriver tests

uranusjr committed:
Do not reuse DAG and task ID in tests

uranusjr committed:
Fix BigQuery tests

uranusjr committed:
Fix log view tests

uranusjr committed:
Fix parameter name in BigQuery tests

uranusjr committed:
Fix (most of) task and tree view tests

uranusjr committed:
Fix Elasticsearch provider tests

uranusjr committed:
Fix Stackdriver test arguments

uranusjr committed:
Fix Google Cloud provider tests

uranusjr committed:
Better DAG IDs

uranusjr committed:
Fix Google MLEngine tests

uranusjr committed:
Fix wasb tests

uranusjr committed:
Fix HTTP sensor tests

uranusjr committed:
Fix Azure Data Explorer tests

uranusjr committed:
Fix Qubole provider tests

uranusjr committed:
Fix SFTP provider tests

uranusjr committed:
Fix SSH provider tests

uranusjr committed:
Fix orphan state reset test session handling

uranusjr committed:
Increase query count for TaskFail.execution_date

uranusjr committed:
Fix scheduler job test session management

uranusjr committed:
Fix trigger rule dep session management

uranusjr committed:
Fix model declaration to match migration

uranusjr committed:
Fix BigQuery test execution_date matching

uranusjr committed:
Fix Dataproc test execution_date matching

uranusjr committed:
Fix MLEngine test execution_date matching

uranusjr committed:
Fix (most of) external task sensor tests

uranusjr committed:
Fix smart sensor operator tests

uranusjr committed:
Fix TI usage in failure callbacks

uranusjr committed:
Fix fixture declaration

ashb committed:
Fix scheduler_job tests

uranusjr committed:
Fix task_stat view execution_date usage

ashb committed:
Fix external task sensors tests

ashb committed:
Fix "Other" tests

Hopefully

ashb committed:
Support filtering by "execution_date" in FAB views

uranusjr committed:
Fix ti attribute loading in /task view

uranusjr committed:
Fix extra link and rendered-templates view tests

ashb committed:
Fix last failing test in www

ashb committed:
Fix migrations for non-Postgres DBs

uranusjr committed:
Fix TI.execution_date in DAG._get_task_instances()

uranusjr committed:
Rewrite gantt view queries to use DagRun join

uranusjr committed:
Further simplify queries in gantt view

uranusjr committed:
Make tests more robust against ti order in db

uranusjr committed:
Less queries on MSSQL!

uranusjr committed:
Fix unique constraint violation in tests

ashb committed:
Fix constraint names on SQLite

uranusjr committed:
Fix scheduler job test by requiring serialized DAG

ashb committed:
fixup! Fix constraint names on SQLite

uranusjr committed:
Attempt to restore session sanity

ashb committed:
fixup! Fix (most of) task and tree view tests

jedcunningham commented on UPDATING.md:
```diff
@@ -228,6 +228,14 @@ Now that the DAG parser syncs DAG permissions there is no longer a need for manu
+In Airflow 2.2 we have changed this and now there is a database-level foreign key constraint ensuring that every TaskInstance has a DagRun row.
+
+Before updating to this 2.2 release you will have to manually resolve any inconsistencies (add back DagRun rows, or delete TaskInstances) if you have any "dangling" TaskInstance" rows.
+
```
Should we also mention that `clean_tis_without_dagrun_interval` has been removed?

ashb replied:
Probably should yes (though having "extra" config items doesn't cause any harm)

ashb committed:
Fix sentry tests

uranusjr committed:
More ti ordering fixes

ashb committed:
Fix kube executor

ashb committed:
Fix KubernetesPodOperator tests

uranusjr commented on airflow/dag_processing/processor.py:
```diff
@@ -387,14 +389,10 @@ def manage_slas(self, dag: DAG, session: Session = None) -> None:
+            TI.dag_id == dag.dag_id,
+            TI.task_id == qry.c.task_id,
+            TI.execution_date == qry.c.max_ti,
         )
```
```suggestion
        max_tis: Iterator[TI] = (
            session.query(TI)
            .join(TI.dag_run)
            .filter(
                TI.dag_id == dag.dag_id,
                TI.task_id == qry.c.task_id,
                DR.execution_date == qry.c.max_ti,
            )
       )
```

I think?

uranusjr commented on airflow/jobs/backfill_job.py:
```diff
@@ -434,15 +433,15 @@ def _process_backfill_task_instances(
-                )
+                ignore_depends_on_past = self.ignore_first_depends_on_past and ti.get_dagrun(
+                    session
+                ).execution_date == (start_date or ti.start_date)
```
```suggestion
                dagrun = self.ignore_first_depends_on_past and ti.get_dagrun(session=session)
                ignore_depends_on_past = dagrun.execution_date == (start_date or ti.start_date)
```

Gosh that Black formatting is awful.

ashb committed:
PR comments

uranusjr committed:
Fix botched suggestion

uranusjr reviewed (approved):
Nervously
